// Code generated by protoc-gen-validate. DO NOT EDIT.
// source: control/v1/control.proto

package controlv1

import (
	"bytes"
	"errors"
	"fmt"
	"net"
	"net/mail"
	"net/url"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode/utf8"

	"google.golang.org/protobuf/types/known/anypb"

	commonv1 "github.com/nanzhong/tstr/api/common/v1"
)

// ensure the imports are used
var (
	_ = bytes.MinRead
	_ = errors.New("")
	_ = fmt.Print
	_ = utf8.UTFMax
	_ = (*regexp.Regexp)(nil)
	_ = (*strings.Reader)(nil)
	_ = net.IPv4len
	_ = time.Duration(0)
	_ = (*url.URL)(nil)
	_ = (*mail.Address)(nil)
	_ = anypb.Any{}
	_ = sort.Sort

	_ = commonv1.Run_Result(0)
)

// Validate checks the field values on RegisterTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RegisterTestRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RegisterTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RegisterTestRequestMultiError, or nil if none found.
func (m *RegisterTestRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *RegisterTestRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if l := utf8.RuneCountInString(m.GetName()); l < 1 || l > 200 {
		err := RegisterTestRequestValidationError{
			field:  "Name",
			reason: "value length must be between 1 and 200 runes, inclusive",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	// no validation rules for Labels

	if m.GetRunConfig() == nil {
		err := RegisterTestRequestValidationError{
			field:  "RunConfig",
			reason: "value is required",
		}
		if !all {
			return err
		}
		errors = append(errors, err)
	}

	if all {
		switch v := interface{}(m.GetRunConfig()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RegisterTestRequestValidationError{
					field:  "RunConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RegisterTestRequestValidationError{
					field:  "RunConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRunConfig()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RegisterTestRequestValidationError{
				field:  "RunConfig",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for CronSchedule

	if all {
		switch v := interface{}(m.GetMatrix()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RegisterTestRequestValidationError{
					field:  "Matrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RegisterTestRequestValidationError{
					field:  "Matrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMatrix()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RegisterTestRequestValidationError{
				field:  "Matrix",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return RegisterTestRequestMultiError(errors)
	}

	return nil
}

// RegisterTestRequestMultiError is an error wrapping multiple validation
// errors returned by RegisterTestRequest.ValidateAll() if the designated
// constraints aren't met.
type RegisterTestRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RegisterTestRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RegisterTestRequestMultiError) AllErrors() []error { return m }

// RegisterTestRequestValidationError is the validation error returned by
// RegisterTestRequest.Validate if the designated constraints aren't met.
type RegisterTestRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RegisterTestRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RegisterTestRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RegisterTestRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RegisterTestRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RegisterTestRequestValidationError) ErrorName() string {
	return "RegisterTestRequestValidationError"
}

// Error satisfies the builtin error interface
func (e RegisterTestRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRegisterTestRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RegisterTestRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RegisterTestRequestValidationError{}

// Validate checks the field values on RegisterTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *RegisterTestResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on RegisterTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// RegisterTestResponseMultiError, or nil if none found.
func (m *RegisterTestResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *RegisterTestResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTest()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, RegisterTestResponseValidationError{
					field:  "Test",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, RegisterTestResponseValidationError{
					field:  "Test",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTest()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return RegisterTestResponseValidationError{
				field:  "Test",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return RegisterTestResponseMultiError(errors)
	}

	return nil
}

// RegisterTestResponseMultiError is an error wrapping multiple validation
// errors returned by RegisterTestResponse.ValidateAll() if the designated
// constraints aren't met.
type RegisterTestResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m RegisterTestResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m RegisterTestResponseMultiError) AllErrors() []error { return m }

// RegisterTestResponseValidationError is the validation error returned by
// RegisterTestResponse.Validate if the designated constraints aren't met.
type RegisterTestResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e RegisterTestResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e RegisterTestResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e RegisterTestResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e RegisterTestResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e RegisterTestResponseValidationError) ErrorName() string {
	return "RegisterTestResponseValidationError"
}

// Error satisfies the builtin error interface
func (e RegisterTestResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sRegisterTestResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = RegisterTestResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = RegisterTestResponseValidationError{}

// Validate checks the field values on UpdateTestRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *UpdateTestRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTestRequestMultiError, or nil if none found.
func (m *UpdateTestRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTestRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFieldMask()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "FieldMask",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "FieldMask",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFieldMask()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateTestRequestValidationError{
				field:  "FieldMask",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Id

	// no validation rules for Name

	if all {
		switch v := interface{}(m.GetRunConfig()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "RunConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "RunConfig",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRunConfig()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateTestRequestValidationError{
				field:  "RunConfig",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Labels

	// no validation rules for CronSchedule

	if all {
		switch v := interface{}(m.GetMatrix()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "Matrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateTestRequestValidationError{
					field:  "Matrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetMatrix()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateTestRequestValidationError{
				field:  "Matrix",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return UpdateTestRequestMultiError(errors)
	}

	return nil
}

// UpdateTestRequestMultiError is an error wrapping multiple validation errors
// returned by UpdateTestRequest.ValidateAll() if the designated constraints
// aren't met.
type UpdateTestRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTestRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTestRequestMultiError) AllErrors() []error { return m }

// UpdateTestRequestValidationError is the validation error returned by
// UpdateTestRequest.Validate if the designated constraints aren't met.
type UpdateTestRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTestRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTestRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTestRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTestRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTestRequestValidationError) ErrorName() string {
	return "UpdateTestRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTestRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTestRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTestRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTestRequestValidationError{}

// Validate checks the field values on UpdateTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateTestResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTestResponseMultiError, or nil if none found.
func (m *UpdateTestResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTestResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return UpdateTestResponseMultiError(errors)
	}

	return nil
}

// UpdateTestResponseMultiError is an error wrapping multiple validation errors
// returned by UpdateTestResponse.ValidateAll() if the designated constraints
// aren't met.
type UpdateTestResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTestResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTestResponseMultiError) AllErrors() []error { return m }

// UpdateTestResponseValidationError is the validation error returned by
// UpdateTestResponse.Validate if the designated constraints aren't met.
type UpdateTestResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTestResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTestResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTestResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTestResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTestResponseValidationError) ErrorName() string {
	return "UpdateTestResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTestResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTestResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTestResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTestResponseValidationError{}

// Validate checks the field values on GetTestRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GetTestRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetTestRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GetTestRequestMultiError,
// or nil if none found.
func (m *GetTestRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetTestRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetTestRequestMultiError(errors)
	}

	return nil
}

// GetTestRequestMultiError is an error wrapping multiple validation errors
// returned by GetTestRequest.ValidateAll() if the designated constraints
// aren't met.
type GetTestRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetTestRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetTestRequestMultiError) AllErrors() []error { return m }

// GetTestRequestValidationError is the validation error returned by
// GetTestRequest.Validate if the designated constraints aren't met.
type GetTestRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetTestRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetTestRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetTestRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetTestRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetTestRequestValidationError) ErrorName() string { return "GetTestRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetTestRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetTestRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetTestRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetTestRequestValidationError{}

// Validate checks the field values on GetTestResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetTestResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetTestResponseMultiError, or nil if none found.
func (m *GetTestResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetTestResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTest()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetTestResponseValidationError{
					field:  "Test",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetTestResponseValidationError{
					field:  "Test",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTest()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetTestResponseValidationError{
				field:  "Test",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetTestResponseMultiError(errors)
	}

	return nil
}

// GetTestResponseMultiError is an error wrapping multiple validation errors
// returned by GetTestResponse.ValidateAll() if the designated constraints
// aren't met.
type GetTestResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetTestResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetTestResponseMultiError) AllErrors() []error { return m }

// GetTestResponseValidationError is the validation error returned by
// GetTestResponse.Validate if the designated constraints aren't met.
type GetTestResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetTestResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetTestResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetTestResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetTestResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetTestResponseValidationError) ErrorName() string { return "GetTestResponseValidationError" }

// Error satisfies the builtin error interface
func (e GetTestResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetTestResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetTestResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetTestResponseValidationError{}

// Validate checks the field values on ListTestsRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListTestsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListTestsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListTestsRequestMultiError, or nil if none found.
func (m *ListTestsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListTestsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListTestsRequestMultiError(errors)
	}

	return nil
}

// ListTestsRequestMultiError is an error wrapping multiple validation errors
// returned by ListTestsRequest.ValidateAll() if the designated constraints
// aren't met.
type ListTestsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListTestsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListTestsRequestMultiError) AllErrors() []error { return m }

// ListTestsRequestValidationError is the validation error returned by
// ListTestsRequest.Validate if the designated constraints aren't met.
type ListTestsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListTestsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListTestsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListTestsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListTestsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListTestsRequestValidationError) ErrorName() string { return "ListTestsRequestValidationError" }

// Error satisfies the builtin error interface
func (e ListTestsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListTestsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListTestsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListTestsRequestValidationError{}

// Validate checks the field values on ListTestsResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListTestsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListTestsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListTestsResponseMultiError, or nil if none found.
func (m *ListTestsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListTestsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetTests() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListTestsResponseValidationError{
						field:  fmt.Sprintf("Tests[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListTestsResponseValidationError{
						field:  fmt.Sprintf("Tests[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListTestsResponseValidationError{
					field:  fmt.Sprintf("Tests[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListTestsResponseMultiError(errors)
	}

	return nil
}

// ListTestsResponseMultiError is an error wrapping multiple validation errors
// returned by ListTestsResponse.ValidateAll() if the designated constraints
// aren't met.
type ListTestsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListTestsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListTestsResponseMultiError) AllErrors() []error { return m }

// ListTestsResponseValidationError is the validation error returned by
// ListTestsResponse.Validate if the designated constraints aren't met.
type ListTestsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListTestsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListTestsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListTestsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListTestsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListTestsResponseValidationError) ErrorName() string {
	return "ListTestsResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListTestsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListTestsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListTestsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListTestsResponseValidationError{}

// Validate checks the field values on ArchiveTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ArchiveTestRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ArchiveTestRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ArchiveTestRequestMultiError, or nil if none found.
func (m *ArchiveTestRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ArchiveTestRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ArchiveTestRequestMultiError(errors)
	}

	return nil
}

// ArchiveTestRequestMultiError is an error wrapping multiple validation errors
// returned by ArchiveTestRequest.ValidateAll() if the designated constraints
// aren't met.
type ArchiveTestRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ArchiveTestRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ArchiveTestRequestMultiError) AllErrors() []error { return m }

// ArchiveTestRequestValidationError is the validation error returned by
// ArchiveTestRequest.Validate if the designated constraints aren't met.
type ArchiveTestRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ArchiveTestRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ArchiveTestRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ArchiveTestRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ArchiveTestRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ArchiveTestRequestValidationError) ErrorName() string {
	return "ArchiveTestRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ArchiveTestRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sArchiveTestRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ArchiveTestRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ArchiveTestRequestValidationError{}

// Validate checks the field values on ArchiveTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ArchiveTestResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ArchiveTestResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ArchiveTestResponseMultiError, or nil if none found.
func (m *ArchiveTestResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ArchiveTestResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ArchiveTestResponseMultiError(errors)
	}

	return nil
}

// ArchiveTestResponseMultiError is an error wrapping multiple validation
// errors returned by ArchiveTestResponse.ValidateAll() if the designated
// constraints aren't met.
type ArchiveTestResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ArchiveTestResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ArchiveTestResponseMultiError) AllErrors() []error { return m }

// ArchiveTestResponseValidationError is the validation error returned by
// ArchiveTestResponse.Validate if the designated constraints aren't met.
type ArchiveTestResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ArchiveTestResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ArchiveTestResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ArchiveTestResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ArchiveTestResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ArchiveTestResponseValidationError) ErrorName() string {
	return "ArchiveTestResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ArchiveTestResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sArchiveTestResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ArchiveTestResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ArchiveTestResponseValidationError{}

// Validate checks the field values on DefineTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DefineTestSuiteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DefineTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DefineTestSuiteRequestMultiError, or nil if none found.
func (m *DefineTestSuiteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *DefineTestSuiteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Name

	// no validation rules for Labels

	if len(errors) > 0 {
		return DefineTestSuiteRequestMultiError(errors)
	}

	return nil
}

// DefineTestSuiteRequestMultiError is an error wrapping multiple validation
// errors returned by DefineTestSuiteRequest.ValidateAll() if the designated
// constraints aren't met.
type DefineTestSuiteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DefineTestSuiteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DefineTestSuiteRequestMultiError) AllErrors() []error { return m }

// DefineTestSuiteRequestValidationError is the validation error returned by
// DefineTestSuiteRequest.Validate if the designated constraints aren't met.
type DefineTestSuiteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DefineTestSuiteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DefineTestSuiteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DefineTestSuiteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DefineTestSuiteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DefineTestSuiteRequestValidationError) ErrorName() string {
	return "DefineTestSuiteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e DefineTestSuiteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDefineTestSuiteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DefineTestSuiteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DefineTestSuiteRequestValidationError{}

// Validate checks the field values on DefineTestSuiteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *DefineTestSuiteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on DefineTestSuiteResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// DefineTestSuiteResponseMultiError, or nil if none found.
func (m *DefineTestSuiteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *DefineTestSuiteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTestSuite()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, DefineTestSuiteResponseValidationError{
					field:  "TestSuite",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, DefineTestSuiteResponseValidationError{
					field:  "TestSuite",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTestSuite()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return DefineTestSuiteResponseValidationError{
				field:  "TestSuite",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return DefineTestSuiteResponseMultiError(errors)
	}

	return nil
}

// DefineTestSuiteResponseMultiError is an error wrapping multiple validation
// errors returned by DefineTestSuiteResponse.ValidateAll() if the designated
// constraints aren't met.
type DefineTestSuiteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m DefineTestSuiteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m DefineTestSuiteResponseMultiError) AllErrors() []error { return m }

// DefineTestSuiteResponseValidationError is the validation error returned by
// DefineTestSuiteResponse.Validate if the designated constraints aren't met.
type DefineTestSuiteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e DefineTestSuiteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e DefineTestSuiteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e DefineTestSuiteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e DefineTestSuiteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e DefineTestSuiteResponseValidationError) ErrorName() string {
	return "DefineTestSuiteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e DefineTestSuiteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sDefineTestSuiteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = DefineTestSuiteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = DefineTestSuiteResponseValidationError{}

// Validate checks the field values on UpdateTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateTestSuiteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTestSuiteRequestMultiError, or nil if none found.
func (m *UpdateTestSuiteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTestSuiteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetFieldMask()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, UpdateTestSuiteRequestValidationError{
					field:  "FieldMask",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, UpdateTestSuiteRequestValidationError{
					field:  "FieldMask",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFieldMask()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return UpdateTestSuiteRequestValidationError{
				field:  "FieldMask",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	// no validation rules for Id

	// no validation rules for Name

	// no validation rules for Labels

	if len(errors) > 0 {
		return UpdateTestSuiteRequestMultiError(errors)
	}

	return nil
}

// UpdateTestSuiteRequestMultiError is an error wrapping multiple validation
// errors returned by UpdateTestSuiteRequest.ValidateAll() if the designated
// constraints aren't met.
type UpdateTestSuiteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTestSuiteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTestSuiteRequestMultiError) AllErrors() []error { return m }

// UpdateTestSuiteRequestValidationError is the validation error returned by
// UpdateTestSuiteRequest.Validate if the designated constraints aren't met.
type UpdateTestSuiteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTestSuiteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTestSuiteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTestSuiteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTestSuiteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTestSuiteRequestValidationError) ErrorName() string {
	return "UpdateTestSuiteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTestSuiteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTestSuiteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTestSuiteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTestSuiteRequestValidationError{}

// Validate checks the field values on UpdateTestSuiteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *UpdateTestSuiteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on UpdateTestSuiteResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// UpdateTestSuiteResponseMultiError, or nil if none found.
func (m *UpdateTestSuiteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *UpdateTestSuiteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return UpdateTestSuiteResponseMultiError(errors)
	}

	return nil
}

// UpdateTestSuiteResponseMultiError is an error wrapping multiple validation
// errors returned by UpdateTestSuiteResponse.ValidateAll() if the designated
// constraints aren't met.
type UpdateTestSuiteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m UpdateTestSuiteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m UpdateTestSuiteResponseMultiError) AllErrors() []error { return m }

// UpdateTestSuiteResponseValidationError is the validation error returned by
// UpdateTestSuiteResponse.Validate if the designated constraints aren't met.
type UpdateTestSuiteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e UpdateTestSuiteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e UpdateTestSuiteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e UpdateTestSuiteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e UpdateTestSuiteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e UpdateTestSuiteResponseValidationError) ErrorName() string {
	return "UpdateTestSuiteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e UpdateTestSuiteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sUpdateTestSuiteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = UpdateTestSuiteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = UpdateTestSuiteResponseValidationError{}

// Validate checks the field values on GetTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetTestSuiteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetTestSuiteRequestMultiError, or nil if none found.
func (m *GetTestSuiteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetTestSuiteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetTestSuiteRequestMultiError(errors)
	}

	return nil
}

// GetTestSuiteRequestMultiError is an error wrapping multiple validation
// errors returned by GetTestSuiteRequest.ValidateAll() if the designated
// constraints aren't met.
type GetTestSuiteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetTestSuiteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetTestSuiteRequestMultiError) AllErrors() []error { return m }

// GetTestSuiteRequestValidationError is the validation error returned by
// GetTestSuiteRequest.Validate if the designated constraints aren't met.
type GetTestSuiteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetTestSuiteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetTestSuiteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetTestSuiteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetTestSuiteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetTestSuiteRequestValidationError) ErrorName() string {
	return "GetTestSuiteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e GetTestSuiteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetTestSuiteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetTestSuiteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetTestSuiteRequestValidationError{}

// Validate checks the field values on GetTestSuiteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *GetTestSuiteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetTestSuiteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetTestSuiteResponseMultiError, or nil if none found.
func (m *GetTestSuiteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetTestSuiteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetTestSuite()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetTestSuiteResponseValidationError{
					field:  "TestSuite",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetTestSuiteResponseValidationError{
					field:  "TestSuite",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTestSuite()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetTestSuiteResponseValidationError{
				field:  "TestSuite",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetTestSuiteResponseMultiError(errors)
	}

	return nil
}

// GetTestSuiteResponseMultiError is an error wrapping multiple validation
// errors returned by GetTestSuiteResponse.ValidateAll() if the designated
// constraints aren't met.
type GetTestSuiteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetTestSuiteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetTestSuiteResponseMultiError) AllErrors() []error { return m }

// GetTestSuiteResponseValidationError is the validation error returned by
// GetTestSuiteResponse.Validate if the designated constraints aren't met.
type GetTestSuiteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetTestSuiteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetTestSuiteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetTestSuiteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetTestSuiteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetTestSuiteResponseValidationError) ErrorName() string {
	return "GetTestSuiteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetTestSuiteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetTestSuiteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetTestSuiteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetTestSuiteResponseValidationError{}

// Validate checks the field values on ListTestSuitesRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListTestSuitesRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListTestSuitesRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListTestSuitesRequestMultiError, or nil if none found.
func (m *ListTestSuitesRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListTestSuitesRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListTestSuitesRequestMultiError(errors)
	}

	return nil
}

// ListTestSuitesRequestMultiError is an error wrapping multiple validation
// errors returned by ListTestSuitesRequest.ValidateAll() if the designated
// constraints aren't met.
type ListTestSuitesRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListTestSuitesRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListTestSuitesRequestMultiError) AllErrors() []error { return m }

// ListTestSuitesRequestValidationError is the validation error returned by
// ListTestSuitesRequest.Validate if the designated constraints aren't met.
type ListTestSuitesRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListTestSuitesRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListTestSuitesRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListTestSuitesRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListTestSuitesRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListTestSuitesRequestValidationError) ErrorName() string {
	return "ListTestSuitesRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListTestSuitesRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListTestSuitesRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListTestSuitesRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListTestSuitesRequestValidationError{}

// Validate checks the field values on ListTestSuitesResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListTestSuitesResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListTestSuitesResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListTestSuitesResponseMultiError, or nil if none found.
func (m *ListTestSuitesResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListTestSuitesResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetTestSuites() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListTestSuitesResponseValidationError{
						field:  fmt.Sprintf("TestSuites[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListTestSuitesResponseValidationError{
						field:  fmt.Sprintf("TestSuites[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListTestSuitesResponseValidationError{
					field:  fmt.Sprintf("TestSuites[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListTestSuitesResponseMultiError(errors)
	}

	return nil
}

// ListTestSuitesResponseMultiError is an error wrapping multiple validation
// errors returned by ListTestSuitesResponse.ValidateAll() if the designated
// constraints aren't met.
type ListTestSuitesResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListTestSuitesResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListTestSuitesResponseMultiError) AllErrors() []error { return m }

// ListTestSuitesResponseValidationError is the validation error returned by
// ListTestSuitesResponse.Validate if the designated constraints aren't met.
type ListTestSuitesResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListTestSuitesResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListTestSuitesResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListTestSuitesResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListTestSuitesResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListTestSuitesResponseValidationError) ErrorName() string {
	return "ListTestSuitesResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListTestSuitesResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListTestSuitesResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListTestSuitesResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListTestSuitesResponseValidationError{}

// Validate checks the field values on ArchiveTestSuiteRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ArchiveTestSuiteRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ArchiveTestSuiteRequest with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ArchiveTestSuiteRequestMultiError, or nil if none found.
func (m *ArchiveTestSuiteRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ArchiveTestSuiteRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return ArchiveTestSuiteRequestMultiError(errors)
	}

	return nil
}

// ArchiveTestSuiteRequestMultiError is an error wrapping multiple validation
// errors returned by ArchiveTestSuiteRequest.ValidateAll() if the designated
// constraints aren't met.
type ArchiveTestSuiteRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ArchiveTestSuiteRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ArchiveTestSuiteRequestMultiError) AllErrors() []error { return m }

// ArchiveTestSuiteRequestValidationError is the validation error returned by
// ArchiveTestSuiteRequest.Validate if the designated constraints aren't met.
type ArchiveTestSuiteRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ArchiveTestSuiteRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ArchiveTestSuiteRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ArchiveTestSuiteRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ArchiveTestSuiteRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ArchiveTestSuiteRequestValidationError) ErrorName() string {
	return "ArchiveTestSuiteRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ArchiveTestSuiteRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sArchiveTestSuiteRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ArchiveTestSuiteRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ArchiveTestSuiteRequestValidationError{}

// Validate checks the field values on ArchiveTestSuiteResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ArchiveTestSuiteResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ArchiveTestSuiteResponse with the
// rules defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ArchiveTestSuiteResponseMultiError, or nil if none found.
func (m *ArchiveTestSuiteResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ArchiveTestSuiteResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ArchiveTestSuiteResponseMultiError(errors)
	}

	return nil
}

// ArchiveTestSuiteResponseMultiError is an error wrapping multiple validation
// errors returned by ArchiveTestSuiteResponse.ValidateAll() if the designated
// constraints aren't met.
type ArchiveTestSuiteResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ArchiveTestSuiteResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ArchiveTestSuiteResponseMultiError) AllErrors() []error { return m }

// ArchiveTestSuiteResponseValidationError is the validation error returned by
// ArchiveTestSuiteResponse.Validate if the designated constraints aren't met.
type ArchiveTestSuiteResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ArchiveTestSuiteResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ArchiveTestSuiteResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ArchiveTestSuiteResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ArchiveTestSuiteResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ArchiveTestSuiteResponseValidationError) ErrorName() string {
	return "ArchiveTestSuiteResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ArchiveTestSuiteResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sArchiveTestSuiteResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ArchiveTestSuiteResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ArchiveTestSuiteResponseValidationError{}

// Validate checks the field values on GetRunRequest with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GetRunRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetRunRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GetRunRequestMultiError, or
// nil if none found.
func (m *GetRunRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetRunRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetRunRequestMultiError(errors)
	}

	return nil
}

// GetRunRequestMultiError is an error wrapping multiple validation errors
// returned by GetRunRequest.ValidateAll() if the designated constraints
// aren't met.
type GetRunRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetRunRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetRunRequestMultiError) AllErrors() []error { return m }

// GetRunRequestValidationError is the validation error returned by
// GetRunRequest.Validate if the designated constraints aren't met.
type GetRunRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetRunRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetRunRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetRunRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetRunRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetRunRequestValidationError) ErrorName() string { return "GetRunRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetRunRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetRunRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetRunRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetRunRequestValidationError{}

// Validate checks the field values on GetRunResponse with the rules defined in
// the proto definition for this message. If any rules are violated, the first
// error encountered is returned, or nil if there are no violations.
func (m *GetRunResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetRunResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// result is a list of violation errors wrapped in GetRunResponseMultiError,
// or nil if none found.
func (m *GetRunResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetRunResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetRun()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetRunResponseValidationError{
					field:  "Run",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetRunResponseValidationError{
					field:  "Run",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRun()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetRunResponseValidationError{
				field:  "Run",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetRunResponseMultiError(errors)
	}

	return nil
}

// GetRunResponseMultiError is an error wrapping multiple validation errors
// returned by GetRunResponse.ValidateAll() if the designated constraints
// aren't met.
type GetRunResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetRunResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetRunResponseMultiError) AllErrors() []error { return m }

// GetRunResponseValidationError is the validation error returned by
// GetRunResponse.Validate if the designated constraints aren't met.
type GetRunResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetRunResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetRunResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetRunResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetRunResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetRunResponseValidationError) ErrorName() string { return "GetRunResponseValidationError" }

// Error satisfies the builtin error interface
func (e GetRunResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetRunResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetRunResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetRunResponseValidationError{}

// Validate checks the field values on ListRunsRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListRunsRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListRunsRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListRunsRequestMultiError, or nil if none found.
func (m *ListRunsRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListRunsRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetScheduledBefore()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "ScheduledBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "ScheduledBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetScheduledBefore()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "ScheduledBefore",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetScheduledAfter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "ScheduledAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "ScheduledAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetScheduledAfter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "ScheduledAfter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetStartedBefore()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "StartedBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "StartedBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetStartedBefore()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "StartedBefore",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetStartedAfter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "StartedAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "StartedAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetStartedAfter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "StartedAfter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetFinishedBefore()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "FinishedBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "FinishedBefore",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFinishedBefore()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "FinishedBefore",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if all {
		switch v := interface{}(m.GetFinishedAfter()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "FinishedAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ListRunsRequestValidationError{
					field:  "FinishedAfter",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetFinishedAfter()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ListRunsRequestValidationError{
				field:  "FinishedAfter",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ListRunsRequestMultiError(errors)
	}

	return nil
}

// ListRunsRequestMultiError is an error wrapping multiple validation errors
// returned by ListRunsRequest.ValidateAll() if the designated constraints
// aren't met.
type ListRunsRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListRunsRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListRunsRequestMultiError) AllErrors() []error { return m }

// ListRunsRequestValidationError is the validation error returned by
// ListRunsRequest.Validate if the designated constraints aren't met.
type ListRunsRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListRunsRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListRunsRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListRunsRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListRunsRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListRunsRequestValidationError) ErrorName() string { return "ListRunsRequestValidationError" }

// Error satisfies the builtin error interface
func (e ListRunsRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListRunsRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListRunsRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListRunsRequestValidationError{}

// Validate checks the field values on ListRunsResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *ListRunsResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListRunsResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListRunsResponseMultiError, or nil if none found.
func (m *ListRunsResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListRunsResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetRuns() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListRunsResponseValidationError{
						field:  fmt.Sprintf("Runs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListRunsResponseValidationError{
						field:  fmt.Sprintf("Runs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListRunsResponseValidationError{
					field:  fmt.Sprintf("Runs[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListRunsResponseMultiError(errors)
	}

	return nil
}

// ListRunsResponseMultiError is an error wrapping multiple validation errors
// returned by ListRunsResponse.ValidateAll() if the designated constraints
// aren't met.
type ListRunsResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListRunsResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListRunsResponseMultiError) AllErrors() []error { return m }

// ListRunsResponseValidationError is the validation error returned by
// ListRunsResponse.Validate if the designated constraints aren't met.
type ListRunsResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListRunsResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListRunsResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListRunsResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListRunsResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListRunsResponseValidationError) ErrorName() string { return "ListRunsResponseValidationError" }

// Error satisfies the builtin error interface
func (e ListRunsResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListRunsResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListRunsResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListRunsResponseValidationError{}

// Validate checks the field values on ScheduleRunRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ScheduleRunRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ScheduleRunRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ScheduleRunRequestMultiError, or nil if none found.
func (m *ScheduleRunRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ScheduleRunRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for TestId

	// no validation rules for Labels

	if all {
		switch v := interface{}(m.GetTestMatrix()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, ScheduleRunRequestValidationError{
					field:  "TestMatrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, ScheduleRunRequestValidationError{
					field:  "TestMatrix",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetTestMatrix()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return ScheduleRunRequestValidationError{
				field:  "TestMatrix",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return ScheduleRunRequestMultiError(errors)
	}

	return nil
}

// ScheduleRunRequestMultiError is an error wrapping multiple validation errors
// returned by ScheduleRunRequest.ValidateAll() if the designated constraints
// aren't met.
type ScheduleRunRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ScheduleRunRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ScheduleRunRequestMultiError) AllErrors() []error { return m }

// ScheduleRunRequestValidationError is the validation error returned by
// ScheduleRunRequest.Validate if the designated constraints aren't met.
type ScheduleRunRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ScheduleRunRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ScheduleRunRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ScheduleRunRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ScheduleRunRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ScheduleRunRequestValidationError) ErrorName() string {
	return "ScheduleRunRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ScheduleRunRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sScheduleRunRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ScheduleRunRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ScheduleRunRequestValidationError{}

// Validate checks the field values on ScheduleRunResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ScheduleRunResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ScheduleRunResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ScheduleRunResponseMultiError, or nil if none found.
func (m *ScheduleRunResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ScheduleRunResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetRuns() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ScheduleRunResponseValidationError{
						field:  fmt.Sprintf("Runs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ScheduleRunResponseValidationError{
						field:  fmt.Sprintf("Runs[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ScheduleRunResponseValidationError{
					field:  fmt.Sprintf("Runs[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ScheduleRunResponseMultiError(errors)
	}

	return nil
}

// ScheduleRunResponseMultiError is an error wrapping multiple validation
// errors returned by ScheduleRunResponse.ValidateAll() if the designated
// constraints aren't met.
type ScheduleRunResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ScheduleRunResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ScheduleRunResponseMultiError) AllErrors() []error { return m }

// ScheduleRunResponseValidationError is the validation error returned by
// ScheduleRunResponse.Validate if the designated constraints aren't met.
type ScheduleRunResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ScheduleRunResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ScheduleRunResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ScheduleRunResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ScheduleRunResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ScheduleRunResponseValidationError) ErrorName() string {
	return "ScheduleRunResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ScheduleRunResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sScheduleRunResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ScheduleRunResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ScheduleRunResponseValidationError{}

// Validate checks the field values on GetRunnerRequest with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetRunnerRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetRunnerRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetRunnerRequestMultiError, or nil if none found.
func (m *GetRunnerRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *GetRunnerRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	// no validation rules for Id

	if len(errors) > 0 {
		return GetRunnerRequestMultiError(errors)
	}

	return nil
}

// GetRunnerRequestMultiError is an error wrapping multiple validation errors
// returned by GetRunnerRequest.ValidateAll() if the designated constraints
// aren't met.
type GetRunnerRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetRunnerRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetRunnerRequestMultiError) AllErrors() []error { return m }

// GetRunnerRequestValidationError is the validation error returned by
// GetRunnerRequest.Validate if the designated constraints aren't met.
type GetRunnerRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetRunnerRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetRunnerRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetRunnerRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetRunnerRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetRunnerRequestValidationError) ErrorName() string { return "GetRunnerRequestValidationError" }

// Error satisfies the builtin error interface
func (e GetRunnerRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetRunnerRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetRunnerRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetRunnerRequestValidationError{}

// Validate checks the field values on GetRunnerResponse with the rules defined
// in the proto definition for this message. If any rules are violated, the
// first error encountered is returned, or nil if there are no violations.
func (m *GetRunnerResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on GetRunnerResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// GetRunnerResponseMultiError, or nil if none found.
func (m *GetRunnerResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *GetRunnerResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if all {
		switch v := interface{}(m.GetRunner()).(type) {
		case interface{ ValidateAll() error }:
			if err := v.ValidateAll(); err != nil {
				errors = append(errors, GetRunnerResponseValidationError{
					field:  "Runner",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		case interface{ Validate() error }:
			if err := v.Validate(); err != nil {
				errors = append(errors, GetRunnerResponseValidationError{
					field:  "Runner",
					reason: "embedded message failed validation",
					cause:  err,
				})
			}
		}
	} else if v, ok := interface{}(m.GetRunner()).(interface{ Validate() error }); ok {
		if err := v.Validate(); err != nil {
			return GetRunnerResponseValidationError{
				field:  "Runner",
				reason: "embedded message failed validation",
				cause:  err,
			}
		}
	}

	if len(errors) > 0 {
		return GetRunnerResponseMultiError(errors)
	}

	return nil
}

// GetRunnerResponseMultiError is an error wrapping multiple validation errors
// returned by GetRunnerResponse.ValidateAll() if the designated constraints
// aren't met.
type GetRunnerResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m GetRunnerResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m GetRunnerResponseMultiError) AllErrors() []error { return m }

// GetRunnerResponseValidationError is the validation error returned by
// GetRunnerResponse.Validate if the designated constraints aren't met.
type GetRunnerResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e GetRunnerResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e GetRunnerResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e GetRunnerResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e GetRunnerResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e GetRunnerResponseValidationError) ErrorName() string {
	return "GetRunnerResponseValidationError"
}

// Error satisfies the builtin error interface
func (e GetRunnerResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sGetRunnerResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = GetRunnerResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = GetRunnerResponseValidationError{}

// Validate checks the field values on ListRunnersRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListRunnersRequest) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListRunnersRequest with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListRunnersRequestMultiError, or nil if none found.
func (m *ListRunnersRequest) ValidateAll() error {
	return m.validate(true)
}

func (m *ListRunnersRequest) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	if len(errors) > 0 {
		return ListRunnersRequestMultiError(errors)
	}

	return nil
}

// ListRunnersRequestMultiError is an error wrapping multiple validation errors
// returned by ListRunnersRequest.ValidateAll() if the designated constraints
// aren't met.
type ListRunnersRequestMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListRunnersRequestMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListRunnersRequestMultiError) AllErrors() []error { return m }

// ListRunnersRequestValidationError is the validation error returned by
// ListRunnersRequest.Validate if the designated constraints aren't met.
type ListRunnersRequestValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListRunnersRequestValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListRunnersRequestValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListRunnersRequestValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListRunnersRequestValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListRunnersRequestValidationError) ErrorName() string {
	return "ListRunnersRequestValidationError"
}

// Error satisfies the builtin error interface
func (e ListRunnersRequestValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListRunnersRequest.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListRunnersRequestValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListRunnersRequestValidationError{}

// Validate checks the field values on ListRunnersResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the first error encountered is returned, or nil if there are no violations.
func (m *ListRunnersResponse) Validate() error {
	return m.validate(false)
}

// ValidateAll checks the field values on ListRunnersResponse with the rules
// defined in the proto definition for this message. If any rules are
// violated, the result is a list of violation errors wrapped in
// ListRunnersResponseMultiError, or nil if none found.
func (m *ListRunnersResponse) ValidateAll() error {
	return m.validate(true)
}

func (m *ListRunnersResponse) validate(all bool) error {
	if m == nil {
		return nil
	}

	var errors []error

	for idx, item := range m.GetRunners() {
		_, _ = idx, item

		if all {
			switch v := interface{}(item).(type) {
			case interface{ ValidateAll() error }:
				if err := v.ValidateAll(); err != nil {
					errors = append(errors, ListRunnersResponseValidationError{
						field:  fmt.Sprintf("Runners[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			case interface{ Validate() error }:
				if err := v.Validate(); err != nil {
					errors = append(errors, ListRunnersResponseValidationError{
						field:  fmt.Sprintf("Runners[%v]", idx),
						reason: "embedded message failed validation",
						cause:  err,
					})
				}
			}
		} else if v, ok := interface{}(item).(interface{ Validate() error }); ok {
			if err := v.Validate(); err != nil {
				return ListRunnersResponseValidationError{
					field:  fmt.Sprintf("Runners[%v]", idx),
					reason: "embedded message failed validation",
					cause:  err,
				}
			}
		}

	}

	if len(errors) > 0 {
		return ListRunnersResponseMultiError(errors)
	}

	return nil
}

// ListRunnersResponseMultiError is an error wrapping multiple validation
// errors returned by ListRunnersResponse.ValidateAll() if the designated
// constraints aren't met.
type ListRunnersResponseMultiError []error

// Error returns a concatenation of all the error messages it wraps.
func (m ListRunnersResponseMultiError) Error() string {
	var msgs []string
	for _, err := range m {
		msgs = append(msgs, err.Error())
	}
	return strings.Join(msgs, "; ")
}

// AllErrors returns a list of validation violation errors.
func (m ListRunnersResponseMultiError) AllErrors() []error { return m }

// ListRunnersResponseValidationError is the validation error returned by
// ListRunnersResponse.Validate if the designated constraints aren't met.
type ListRunnersResponseValidationError struct {
	field  string
	reason string
	cause  error
	key    bool
}

// Field function returns field value.
func (e ListRunnersResponseValidationError) Field() string { return e.field }

// Reason function returns reason value.
func (e ListRunnersResponseValidationError) Reason() string { return e.reason }

// Cause function returns cause value.
func (e ListRunnersResponseValidationError) Cause() error { return e.cause }

// Key function returns key value.
func (e ListRunnersResponseValidationError) Key() bool { return e.key }

// ErrorName returns error name.
func (e ListRunnersResponseValidationError) ErrorName() string {
	return "ListRunnersResponseValidationError"
}

// Error satisfies the builtin error interface
func (e ListRunnersResponseValidationError) Error() string {
	cause := ""
	if e.cause != nil {
		cause = fmt.Sprintf(" | caused by: %v", e.cause)
	}

	key := ""
	if e.key {
		key = "key for "
	}

	return fmt.Sprintf(
		"invalid %sListRunnersResponse.%s: %s%s",
		key,
		e.field,
		e.reason,
		cause)
}

var _ error = ListRunnersResponseValidationError{}

var _ interface {
	Field() string
	Reason() string
	Key() bool
	Cause() error
	ErrorName() string
} = ListRunnersResponseValidationError{}
